app/controllers/mass_rollouts_controller.rb

```diff
@@ -16,23 +16,37 @@ def create
   end
 
   def deploy
-    environment = deploy_group.environment
-    template_stages = environment.template_stages.all
-    missing_only = params[:missing_only] == "true"
-    stages_to_deploy = missing_only ? deploy_group.stages.reject(&:last_successful_deploy) : deploy_group.stages
-    deploys = stages_to_deploy.map do |stage|
-      template_stage = template_stages.detect { |ts| ts.project_id == stage.project.id }
-      next unless template_stage
+    successful, missing = deploy_group.stages.partition(&:last_successful_deploy)
+    stages_to_deploy = []
+
+    if params[:successful] == "true"
+      stages_to_deploy += successful.map { |stage| [stage, stage.last_successful_deploy.reference] }
+    end
 
-      last_success_deploy = template_stage.last_successful_deploy
-      next unless last_success_deploy
+    if params[:missing] == "true"
+      stages_to_deploy += missing.map do |stage|
+        ref = stage.template_stage.last_successful_deploy&.reference if stage.template_stage&.is_template?
```

Comment:
note this behaviour is slightly different.. it'll use the actual template stage referenced under the stage **NOT** just the first template stage that matches the project ID and environment. not sure is a breaking change or not. /cc @zendesk/bre